components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpoint.java

```diff
@@ -67,202 +71,188 @@
-                    }
+        throw new TokenEndpointBadRequestException(null, null, e);
+    }
 
-                    if (!OAuthConstants.OauthAppStates.APP_STATE_ACTIVE.equalsIgnoreCase(appState)) {
-                        if (log.isDebugEnabled()) {
-                            log.debug("Oauth App is not in active state.");
-                        }
-                        OAuthResponse oAuthResponse = OAuthASResponse.errorResponse(HttpServletResponse
-                                .SC_UNAUTHORIZED)
-                                .setError(OAuth2ErrorCodes.INVALID_CLIENT)
-                                .setErrorDescription("Oauth application is not in active state.")
-                                .buildJSONMessage();
-                        return Response.status(oAuthResponse.getResponseStatus()).entity(oAuthResponse.getBody())
-                                .build();
-                    }
-                } catch (IdentityOAuthAdminException e) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Error in getting oauth app state.", e);
-                    }
-                    OAuthResponse oAuthResponse = OAuthASResponse.errorResponse(HttpServletResponse.SC_NOT_FOUND)
-                            .setError(OAuth2ErrorCodes.SERVER_ERROR)
-                            .setErrorDescription("Error in getting oauth app state.").buildJSONMessage();
-                    return Response.status(oAuthResponse.getResponseStatus()).entity(oAuthResponse.getBody())
-                            .build();
+    private void validateRepeatedParams(HttpServletRequest request, MultivaluedMap<String, String> paramMap)
+            throws TokenEndpointBadRequestException {
+
+        if (!validateParams(request, paramMap)) {
+            throw new TokenEndpointBadRequestException("Invalid request with repeated parameters.");
+        }
+    }
+
+    private void validateOAuthApplication(HttpServletRequestWrapper httpRequest) throws InvalidApplicationClientException,
+            TokenEndpointBadRequestException {
+
+        String consumerKey = getConsumerKey(httpRequest);
+
+        if (isNotEmpty(consumerKey)) {
+            validateOauthApplication(consumerKey);
+        } else {
+            throw new TokenEndpointBadRequestException("Missing parameters: client_id");
+        }
+    }
+
+    private Response buildTokenResponse(OAuth2AccessTokenRespDTO oauth2AccessTokenResp) throws OAuthSystemException {
+
+        OAuthTokenResponseBuilder oAuthRespBuilder = OAuthASResponse
+                .tokenResponse(HttpServletResponse.SC_OK)
+                .setAccessToken(oauth2AccessTokenResp.getAccessToken())
+                .setRefreshToken(oauth2AccessTokenResp.getRefreshToken())
+                .setExpiresIn(Long.toString(oauth2AccessTokenResp.getExpiresIn()))
+                .setTokenType(BEARER);
+        oAuthRespBuilder.setScope(oauth2AccessTokenResp.getAuthorizedScopes());
```

Comment:
Chain this call as done above

Reply:
Not possible to chain setScope method since it does not return OAuthTokenResponseBuilder